plugins/plugin-logui/src/controller/kubectl/logs.ts

```diff
@@ -14,102 +14,160 @@
-
-  if (!streamed && (hasResource || hasSelector)) {
-    // render as Kui table
-    return getLogsAsTable(args)
-  } else {
-    // send to PTY
-    if (streamed && !args.parsedOptions.since) {
-      // see https://github.com/kui-shell/plugin-kubeui/issues/210
-      const since = '10s'
-      args.parsedOptions.since = since
-      args.argvNoOptions.push(since)
-      args.argv.push(since)
-      args.command = args.command + ' --since=10s'
-    }
 
-    const stdout = await args.createOutputStream()
-    const myExecOptions = Object.assign({}, args.execOptions, {
-      quiet: true,
-      replSilence: true,
-      echo: false,
-      onInit: (ptyJob: Abortable) => {
-        return _ => {
-          if (args.block['isCancelled']) {
-            ptyJob.abort()
+  // if we are streaming (logs -f), and the user did not specify a
+  // "--since", then add one, to prevent the default behavior of
+  // kubectl which fetches quite a bit of history
+  if (streamed && !args.parsedOptions.since) {
+    // see https://github.com/kui-shell/plugin-kubeui/issues/210
+    const since = '10s'
+    args.parsedOptions.since = since
+    args.argv.push('--since=since')
+    args.command = args.command + ' --since=10s'
+  }
+
+  // if we are not streaming, and the user has not specified a
+  // "--tail", then add one, for the same reason
+  if (!streamed && !args.parsedOptions.tail) {
+    const tail = 30
+    args.parsedOptions.tail = tail
+    args.argv.push('--tail=' + tail)
+    args.command = args.command + ' --tail=' + tail
+  }
+
+  // set up the PTY stream; we want to stream to this stdout sink
+  const stdout = await args.createOutputStream()
+
+  // a bit of plumbing: tell the PTY that we will be handling everything
+  const myExecOptions = Object.assign({}, args.execOptions, {
+    rethrowErrors: true, // we want to handle errors
+    quiet: true, // don't ever emit anything on your own
+    replSilence: true, // repl: same thing
+    echo: false, // do not even echo "ok"
+
+    // the PTY will call this when the PTY process is ready; in
+    // return, we send it back a consumer of streaming output
+    onInit: (ptyJob: Abortable) => {
+      let curLine: string
+
+      // _ is one chunk of streaming output
+      return (_: Streamable) => {
+        if (args.block['isCancelled']) {
+          ptyJob.abort()
+        } else if (typeof _ === 'string') {
+          // we only know how to handle strings
+          if (/\n$/.test(_)) {
+            const joined = curLine ? curLine + _ : _
+            // if the output from the PTY already contains ANSI
+            // control characters, then we won't add any of our
+            // own. \x1b is ESCAPE
+            // eslint-disable-next-line no-control-regex
+            const fullLine = /\x1b/.test(joined)
+              ? joined
+              : joined
+                  // informational extras
+                  .replace(/(\[.*?\])/g, (match, p1) => colors.gray(p1))
```

Comment:
These rules looks good, but have some concerns:
1.  Will it be better to wrap all these rules into a function
2. All rules seems targeting to one format. Will it be nice if add some comments with examples of the format?
3. Looks like a lot of replace (11+ global matches without ^&), not sure whether it will slow the system down. 
    Is these code run on the webpack side or host(server/proxy) side? 
    If on webpack side, I think we don't need to worry about it.

Reply:
this code is run on the client, no the proxy server

Reply:
pushed a cleanup: separated out decorateLogLines and added comments and examples!